### ISSUE TITLE
non error when returning a sources object with non-existing state change function

### ISSUE THREAD

**Issue by xfh** [state-adapt/state-adapt#67]
Hi Mike

Do you see a way to make the sources object strictly typed?

I've noticed that there are no compiler errors when a non-existing state change function is declared. That's unfortunate, because renaming a state change function (or a state property when using joinAdapters) quickly results in breaking code.
One has to remember to check all usages of the adapter and update the state change function manually.

In the following example, the count store uses the 4th adapt overload and declares `nonExistingStateChangeFunction`, which should cause a compiler error.
```ts
  source$ = new Source<number>('foo');
  countStore = adapt(0, {
    adapter,
    sources: (store) => {
      return {
        // commenting the following line will fail typescript compilation (as desired)
        set: this.source$,
        // typically, using joinAdapters, you automatically get setMyProperty state change functions.
        // Sometimes a property gets renamed. One has to remember to update the state change function accordingly.
        nonExistingStateChangeFunction: this.source$,
      };
    },
  });
```

I have the same issue with the 3rd overload, not detecting `yetAnotherNonExistingStateChangeFunction`:
```ts
  nameChange$ = new Source<string>('name');
  nameStore = adapt('John', {
    sources: {
      // does not fail either
      yetAnotherNonExistingStateChangeFunction: this.nameChange$,
    },
  });
```

Interestingly, the 4th overload causes an error when the returned source object doesn't contain any legitimate state change functions (remove or rename `set` in the example).

Not sure if it matters, but the example are with adapt from the angular package. 
Here is a stackblitz https://stackblitz.com/edit/state-adapt-angular-vxoxab?file=src%2Fapp%2Fcounter.component.ts

Thanks for looking into this.

**Comment by mfp22** [comment]
These aren't technically `adapt` overloads, so I was confused for a second thinking this was StateAdapt 1.x. These are just different ways to define sources. And I thought when you said "quickly results in breaking code" that it was resulting in a runtime error, but I checked the source code just now to be sure and it doesn't try to call a non-existent function I'm pretty sure. It just doesn't use the source.

Obviously it would be great to get a compiler error, and I know more about TypeScript now than when I wrote these original types, so I'll think about how this could be done... But the reason it doesn't error is because TypeScript sees an object literal and it never enforces that extra properties aren't passed in, as long as the ones it expects are there. Since all the properties are optional, you can pass none of the expected ones and a bunch of random ones and it won't care. Inference probably works differently with the sources function and it just doesn't know what object you're returning perhaps? I'm not sure about that one. 

It's also a problem with state change functions. If you spread new state and add a non-existent property, it won't complain. Here's an example where I've added a `randomProperty` onto some NgRx state. https://stackblitz.com/edit/ngrx-platform-demo-htjfdm?file=example-app%2Fapp%2Fcore%2Freducers%2Flayout.ts

While refactoring other libraries to StateAdapt, I noticed the same thing and it was really annoying. I've run into it for state objects most often. But sources objects as well. It's only when returning object literals from functions, or first defining them before passing them into functions that don't expect the object property, that it allows it. Here's a reference to this behavior: https://stackoverflow.com/questions/68561033/typescript-object-assignment-with-excess-keys-seems-broken 

But, like I said, I'm much better at TypeScript now, so I might be able to figure something out with conditional types and `never`.

**Comment by mfp22** [comment]
Yeah, something like this https://observablehq.com/@koop/excess-property-checking-in-typescript

What do you think, should I add signal support first, or work on this first?

**Comment by mfp22** [comment]
This actually seems to work 
![image](https://github.com/state-adapt/state-adapt/assets/13700926/d99a597e-1288-4fa8-8f70-6eb1fb7af711)

And in your example, as soon as you provide an adapter it works for the sources object argument. So either you only have `set`, `noop`, `update` or `reset`, or you have misspelled one of those and it does allow that. This is not so bad, but still not ideal I guess. 

It seems like the biggest problem is when defining a sources factory function where you're using a custom adapter too. 

So here are the test cases I just added 
![image](https://github.com/state-adapt/state-adapt/assets/13700926/69302b6e-4989-424a-8e7f-2a9e5beb3251)

**Comment by xfh** [comment]
Sorry for the confusion. Not always easy to keep the right terminology. I am glad you got what I was referring to :-)

Thanks for the interesting read. I tried to work on the typings but you beat me to it. I haven't even managed the sources object argument yet. Curious how you did it.

An idea for the sources factory function: replace the arrow function with a utility function, where you can specify the types as in the articles `set` function. Something like:
```ts
sources: watched(store => ({...}))
```
But that would be a breaking change...

**Comment by mfp22** [comment]
Nah, the `watched` function would need the type information from the other arguments passed in...
I'm trying the approach from that link but struggling because I may already have too many generics

**Comment by mfp22** [comment]
Well I forked your repro to make a simple example of the problem that needs to be solved https://stackblitz.com/edit/state-adapt-angular-38pewt?file=src%2Fapp%2Fcounter.component.ts

I'll ask Twitter

**Comment by xfh** [comment]
I spent a day working on this, but didn't succeed. In case it helps, I've found out that the NgRx example with the reducer is detecting unknown properties in newer versions of NgRx: https://stackblitz.com/edit/ngrx-platform-demo-186r2l?file=example-app%2Fapp%2Fcore%2Freducers%2Flayout.ts

The trick might be the usage of InferredState, introduced in this pull request https://github.com/ngrx/platform/pull/3577

**Comment by mfp22** [comment]
I asked Twitter and got the simple case solved, but not when it could be a regular object OR a factory function. I think I'll ask for help again. But in the worst case, I think I could solve this if the factory function were defined in a separate property. So we'd have `sources` and `getSources`. Would that be an okay tradeoff? 

In the actual StateAdapt code it's a little more complicated. For now I think I'll leave some comments in the code and let this be an open issue for someone to help solve.

**Comment by mfp22** [comment]
Asked on Twitter again https://twitter.com/mfpears/status/1756580237725532161

**Comment by mfp22** [comment]
I decided to throw Grok 4.6 at this because I'm an AI sadist apparently and it is having the hardest time 😂 It's been going so long.

<img width="632" height="90" alt="Image" src="https://github.com/user-attachments/assets/0dcda3d6-0362-40d0-88a9-0d8c95390eb5" />

5.6 Sol next

**Comment by mfp22** [comment]
Grok actually did get it, but it was too complicated, so I had 5.6 Sol go over it and simplify it, and it's good. Will release with StateAdapt 4.0. 
7e92484e2f9ddfe00b1cbffcd351537a30653d4a
